specs/applicationdata/ApplicationData.md

```diff
@@ -0,0 +1,428 @@
+
+## 3.4. Per-Machine data store
+
+Package families can optionally have a per-machine data store.
+
+### 3.4.1. Machine Path/Folder
+
+The path for a package family’s "Machine Folder" is located at `%ProgramData%\Microsoft\Windows\AppRepository\ApplicationData\...packagefamilyname...\Machine`.
+
+This directory is ACL’d similarly to a package’s System Metadata directory (`%ProgramData%\Microsoft\Windows\AppRepository\Packages\...pkgfullname...`) e.g.
+
+```
+BUILTIN\Users:(OI)(CI)(Rc,S,RD,REA,X,RA)
+S-1-15-3-2977037414-864741429-1129033548-1928484290-1803339615-1058153653-556172075:(OI)(CI)(RX)
+BUILTIN\Users:(OI)(CI)(R)
+NT SERVICE\TrustedInstaller:(I)(F)
+NT SERVICE\TrustedInstaller:(I)(CI)(IO)(F)
+NT AUTHORITY\SYSTEM:(I)(F)
+NT AUTHORITY\SYSTEM:(I)(OI)(CI)(IO)(F)
+BUILTIN\Administrators:(I)(RX)
+BUILTIN\Administrators:(I)(OI)(CI)(IO)(GR,GE)
+S-1-15-3-1024-3635283841-2530182609-996808640-1887759898-3848208603-3313616867-983405619-2501854204:(I)(RX)
+S-1-15-3-1024-3635283841-2530182609-996808640-1887759898-3848208603-3313616867-983405619-2501854204:(I)(OI)(CI)(IO)(GR,GE)
+```
+
+### 3.4.2. Manifested Opt-In
+
+The "Machine Folder" is only available for packages that explicitly opt-in to enabling it via appxmanifest.xml
+
+```xml
+<Package>
+  ...
+  <Properties>
+    <appdata:ApplicationData>
+      <MachineFolder/>
+    </appdata:ApplicationData>
+    ...
+```
+
+### 3.4.3. Machine Path Creation/Deletion
+
+A new Undocked Deployment Extension Handler (DEH) creates the Machine path when a package in the package family is installed, if it doesn't already exist.
+
+The Machine path is removed when the last package in the package family is removed from the machine.
+
+### 3.4.4. APIs
+
+New properties and methods to access per-machine data:
+
+* ApplicationDataLocality.Machine
+* ClearMachineFolderAsync()
+* IsMachinePathSupported
+* MachineFolder
+* MachinePath
+
+**NOTE:** The per-machine APIs require the caller has the package family registered for the user -or- is running as SYSTEM.
+
+## 3.5. Unpackaged app data stores
+
+Unpackaged applications can use the ApplicationData API to access app data stores in the classic locations:
+
+|Packaged                        |Unpackaged                                                         |
+|--------------------------------|-------------------------------------------------------------------|
+|ApplicationData.LocalCachePath  |%LOCALAPPDATA%\<publisher>\<product>                               |
+|ApplicationData.LocalPath       |%LOCALAPPDATA%\<publisher>\<product>                               |
+|ApplicationData.MachinePath     |%ProgramData%\<publisher>\<product>                                |
+|ApplicationData.RoamingPath     |%APPDATA%                                                          |
+|ApplicationData.TemporaryPath   |[GetTempPath2W()](https://learn.microsoft.com/windows/win32/api/fileapi/nf-fileapi-gettemppath2w)<sup>1</sup>|
```

Comment:
Any idea where we've provided _written_ guidance to unpackaged apps?

"Get Temp dir" (and its various API variations over the years) is a common practice for decades and I recall the "store your data/config under HKCU\SOFTWARE\company\product (or %LOCALAPPDATA%)" guidance when Win95 came out, and cultural norms and common guidance and practice since. But I don't know where it's formally documented these days. Wondering if that says anything about TEMP, and if not if there's any de facto common practice re temp (I've seen both styles - subdir and general mosh pit).

I'm a little hesitant to dictate ApplicationData.GetForUnpackaged("x", "y").TemporaryFolder == GetTempPath2W() + "\x\y" as someone will have to create that directory and, presumably destroy it. We could create it on 1st use but who destroys it? Natural expectation is whoever creates it, but won't have any idea when it's "done" and can be removed, and if the app doesn't create it would they routinely think to remove it? Or do we just crate large swathes of temp files hanging around 'forever' (or if/until someone runs DiskCleanupWizard etc)?

Reply:
I think Storage Sense runs around cleaning up stale files in TEMP directories, as well as cleaning up empty directories. I mean, *somebody* has to. Otherwise, those files just accumulate indefinitely!

Having a practical difference between packaged and unpackaged apps seems wrong. "I upgraded my app to packaged, and stuff doesn't work any more." Or worse "I downgraded my app from packaged to unpackaged, and now my app crashes (because it now encounters sharing violations on files in its TemporaryFolder)."

Reply:
CONSIDER: Unpackaged `TemporaryPath` = `GetTempPath2W()\publisher\product`